lib/src/client.dart

```diff
@@ -151,14 +154,26 @@ class Client extends http.BaseClient {
       throw StateError("$prefix can't be refreshed.");
     }
 
-    _credentials = await credentials.refresh(
-        identifier: identifier,
-        secret: secret,
-        newScopes: newScopes,
-        basicAuth: _basicAuth,
-        httpClient: _httpClient);
-
-    if (_onCredentialsRefreshed != null) _onCredentialsRefreshed(_credentials);
+    // To make sure that only one refresh happens when credentials are expired
+    // we track it using the [_refreshingFuture]. And also make sure that the
+    // _onCredentialsRefreshed callback is only called once.
+    if (_refreshingFuture == null) {
+      try {
+        _refreshingFuture = credentials.refresh(
+          identifier: identifier,
+          secret: secret,
+          newScopes: newScopes,
+          basicAuth: _basicAuth,
+          httpClient: _httpClient,
+        );
+        _credentials = await _refreshingFuture;
+        _onCredentialsRefreshed?.call(_credentials);
+      } finally {
+        _refreshingFuture = null;
+      }
+    } else {
+      await _refreshingFuture;
+    }
 
     return this;
```

Comment:
Nit, but consider:
```dart
return _refreshingFuture ?? Future.microtask(() async {
  try {
    _credentials = await credentials.refresh(....);
    _onCredentialsRefreshed?.call(_credentials);
    return this;
  } finally {
    _refreshingFuture = null;
  }
});
```

That way all the errors gets the same stack trace. Not important, I'm pretty your solution will work just as fine :D

Reply:
Assuming you mean `??=` that’s also a nice solution.
But personally I would actually prefer being able to see the different stack traces.
Then you can track which call actually triggered the refresh, although indeed usually it should not be important.

Reply:
> Then you can track which call actually triggered the refresh, 

Hmm, that's a good point :D